src/matches.ts

```diff
@@ -0,0 +1,56 @@
+import { purry } from './purry';
+
+function _matches(target: any, condition: any) {
+  if (typeof condition === 'function') return condition(target);
+
+  if (typeof condition === 'object') {
+    return Object.entries(condition).every(([key, value]) =>
+      matches(target[key], value)
+    );
+  }
+
+  return target === condition;
+}
+
+type Matcher<I = any> = ((dataInput: I) => boolean) | any;
```

Comment:
[default for generic] For the type at line 52. 

[ | any ] Because a matcher can also be anything. E.g.

```
matches({ id: '123 })(obj);
matches(true)(isValid);
```

And I know that `((dataInput: I) => boolean) | any` could simply be `any`, but a little explicitness for the most used case shouldn't hurt.

Reply:
I agree with @TkDodo, these types don't make much sense because they are only for readability, they don't provide any type *safety* for users. In fact this allows users to send invalid inputs that would fail silently (mainly just returning false for them) without the users ever knowing about it until it explodes in production.

Can we work on the type of Matches a little more?
1. I think we can start with the fact that I should be `object`. This function doesn't make sense for any other use-cases because they can simply use `equals`.
2. Then it should map the props of the input object and recursively allow more mappers for any object values.

Here's a crude first attempt:
```suggestion
type Matcher<T extends object> = {
  [K in keyof T]?: T[K] extends object ? Matcher<T[K]> : ((value: T[K]) => boolean) | T[K]
}
```